### PR TITLE
feat(edge): auto update process

### PR DESCRIPTION
**Pull request by deviantony**
Related to https://github.com/portainer/agent/pull/284 [EE-3023]

[EE-3023]: https://portainer.atlassian.net/browse/EE-3023?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -216,9 +223,83 @@ func main() {
 
 	// !Edge
 
+	if options.HealthCheck {
+		log.Println("[INFO] [healthcheck] [message: Agent running on in healthcheck mode. Will exit after running pre-flight checks.]")
```
Comment (chiptus) [inline cmd/agent/main.go:227]: we may consider moving it to a separate binary

File: cmd/agent/main.go [hunk 2]
```diff
@@ -216,9 +223,83 @@ func main() {
...
+
+		if !options.EdgeMode {
+
+			// TODO: REVIEW
+			// Healthcheck not considered for regular agent in the scope of the agent auto-upgrade POC
```
Comment (chiptus) [inline cmd/agent/main.go:232]: don't see why not

File: cmd/agent/main.go [hunk 3]
```diff
@@ -216,9 +223,83 @@ func main() {
...
+			goos.Exit(0)
+		}
+
+		// Edge pre-flight checklist
+		// Here we check that the values in the Edge Key can be used by the agent
+		// * The agent can reach out the Portainer instance
+		// * HTTP polling is reachable
+		// * Tunneling is reachable
+
+		edgeKey := edgeManager.GetKey()
```
Comment (chiptus) [inline cmd/agent/main.go:244]: edgeKey can be nil (if it wasn't set). I guess this would mean `unhealthy`?

File: cmd/agent/main.go [hunk 4]
```diff
@@ -216,9 +223,83 @@ func main() {
...
+
+		// We check that the Portainer instance URL is valid
```
Comment (chiptus) [inline cmd/agent/main.go:246]: what about checking if the agent api is available? or is that not needed as in that case the process will be closed

Comment (chiptus) [reply cmd/agent/main.go:246]: one case where this might happen is if the API is deadlocked

File: build/linux/Dockerfile [hunk 5]
```diff
@@ -5,4 +5,10 @@ WORKDIR /app
 COPY dist /app/
 COPY static /app/static
 
+# TODO: REVIEW
+# Health check only available on Linux containers for the scope of the POC
```
Comment (chiptus) [inline build/linux/Dockerfile:9]: the only change needed to make it available in other types is copy-pasting L12, right?

File: cmd/agent/main.go (hunk 1 above)
Comment (deviantony) [reply cmd/agent/main.go:227]: I don't have any particular opinion on that.

File: cmd/agent/main.go (hunk 2 above)
Comment (deviantony) [reply cmd/agent/main.go:232]: To be discussed with the product team regarding the scope of the piece of work, this was just stated for the POC scope.

File: cmd/agent/main.go (hunk 4 above)
Comment (deviantony) [reply cmd/agent/main.go:246]: What could cause the agent API to be deadlocked/not avail?

File: cmd/agent/main.go (hunk 3 above)
Comment (deviantony) [reply cmd/agent/main.go:244]: That's a valid point, I did not consider the manual input of an Edge key (via the Edge UI) in that POC. This is something that we'll need to handle properly.

File: build/linux/Dockerfile (hunk 5 above)
Comment (deviantony) [reply build/linux/Dockerfile:9]: I think that we'll need to change the CMD but I never defined a healthcheck on a Windows container

File: cmd/agent/main.go (hunk 4 above)
Comment (chiptus) [reply cmd/agent/main.go:246]: I don't know, but what if :)

**Comment by chiptus** [comment]
closed in favor of work done on [EE-3023]


[EE-3023]: https://portainer.atlassian.net/browse/EE-3023?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Pull request closed without merging**